File: client/web/src/utils/countryContext.js
/**
 * Country Context Utility (Web)
 * Single source of truth for country-to-currency mapping and other defaults.
 */

export const getCountryContext = (countryCode) => {
  // 1. Normalize input
  const code = (countryCode || 'SA').toUpperCase().trim();
  
  // 2. Define the static mapping (Single Source of Truth)
  const mapping = {
    'SA': { 
      currencyCode: 'SAR',
      countryName: 'Saudi Arabia',
      defaultVatRate: 15
    },
    'EG': { 
      currencyCode: 'EGP',
      countryName: 'Egypt',
      defaultVatRate: 14
    },
    'AE': { 
      currencyCode: 'AED',
      countryName: 'Emirates',
      defaultVatRate: 5
    },
    'KW': { 
      currencyCode: 'KWD',
      countryName: 'Kuwait',
      defaultVatRate: 0
    },
    'QA': { 
      currencyCode: 'QAR',
      countryName: 'Qatar',
      defaultVatRate: 0
    }
  };
  
  // 3. Resolve context (Strict lookup)
  const context = mapping[code] || mapping['SA'];
  
  return {
    countryCode: code,
    currencyCode: context.currencyCode,
    countryName: context.countryName,
    defaultVatRate: context.defaultVatRate
  };
};

File: mobile/foodie/lib/utils/country_context.dart
class CountryContext {
  final String countryCode;
  final String currencyCode;
  final String countryName;
  final double defaultVatRate;

  CountryContext({
    required this.countryCode,
    required this.currencyCode,
    required this.countryName,
    required this.defaultVatRate,
  });
}

class CountryContextHelper {
  static CountryContext getContext(String? code) {
    // 1. Normalize input
    final normalizedCode = (code ?? 'SA').toUpperCase().trim();
    
    // 2. Define the static mapping (Single Source of Truth)
    final Map<String, CountryContext> mapping = {
      'SA': CountryContext(
        countryCode: 'SA',
        currencyCode: 'SAR',
        countryName: 'Saudi Arabia',
        defaultVatRate: 15,
      ),
      'EG': CountryContext(
        countryCode: 'EG',
        currencyCode: 'EGP',
        countryName: 'Egypt',
        defaultVatRate: 14,
      ),
      'AE': CountryContext(
        countryCode: 'AE',
        currencyCode: 'AED',
        countryName: 'United Arab Emirates',
        defaultVatRate: 5,
      ),
      'KW': CountryContext(
        countryCode: 'KW',
        currencyCode: 'KWD',
        countryName: 'Kuwait',
        defaultVatRate: 0,
      ),
      'QA': CountryContext(
        countryCode: 'QA',
        currencyCode: 'QAR',
        countryName: 'Qatar',
        defaultVatRate: 0,
      ),
    };

    // 3. Resolve context (Strict lookup with safety fallback)
    return mapping[normalizedCode] ?? CountryContext(
      countryCode: normalizedCode,
      currencyCode: 'USD',
      countryName: normalizedCode,
      defaultVatRate: 0,
    );
  }
}
